File: examples/external_quote_subscriber.js
#!/usr/bin/env node
/**
 * examples/external_quote_subscriber.js
 *
 * EvTrade 行情订阅 Demo (JS / Node)
 * =====================================================================
 *
 * 演示两种订阅模式:
 *   模式 A: 全量订阅 (subscribe_all=true) → 后端 broadcast_to_stock() 对所有
 *            当前有 tick 的标的过滤推送,前端 ws 只要 send {type:"subscribe", stock_codes:[...]}
 *            即按订阅走
 *   模式 B: 按证券代码订阅 (subscribe_all=false + stock_codes=[...]) →
 *            只订阅指定代码,其他 tick 即使来了也收不到
 *
 * 用法:
 *   # 1. 全量模式 (默认订阅几个演示代码,后端只推这些;零订阅时没数据)
 *   node examples/external_quote_subscriber.js
 *
 *   # 2. 按证券代码模式
 *   node examples/external_quote_subscriber.js --mode=specific --codes=000001,600519,000333
 *
 *   # 3. 切换 base URL
 *   BASE_URL=http://192.168.1.100:8000 node examples/external_quote_subscriber.js
 *
 *   # 4. 不订阅任何代码 (验证"零订阅时静默丢弃",看 60 秒有没有数据)
 *   node examples/external_quote_subscriber.js --mode=none
 *
 *   # 5. JWT 调试 (跳过 login)
 *   JWT_TOKEN=eyJhbGc... node examples/external_quote_subscriber.js
 *
 * 依赖: ws (Node.js 库, 项目 client/node_modules 已装)
 * 运行: cd /root/workspcae/codespace/EvTrade && node examples/external_quote_subscriber.js
 *
 * 登录: quota / quota (viewer 角色, 行情 demo 专用账号)
 *
 * 退出: Ctrl+C 优雅关闭 (清理 ws + 退出)
 * =====================================================================
 */

const WebSocket = require("../client/node_modules/ws");

// ────────────────────── 配置区 ──────────────────────
const CONFIG = {
  baseUrl: process.env.BASE_URL || "http://127.0.0.1:8000",
  wsUrl: null, // 根据 baseUrl 推导
  username: process.env.USERNAME || "quota",
  password: process.env.PASSWORD || "quota",
  // 演示用 codes (可在 CLI 覆盖)
  defaultCodes: ["000001", "600519", "000333", "000858", "601318"],
  // 收集时长 (秒) — 收到 N 条 tick 或超时后自动退出,打印统计
  collectSeconds: parseInt(process.env.COLLECT_SECONDS || "12", 10),
  // 打印频率 — 防止刷屏, 默认每个标的第一条 + 每秒心跳
  heartbeatSec: 5,
};

// ────────────────────── CLI 参数解析 ──────────────────────
function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { mode: "specific", codes: CONFIG.defaultCodes };
  for (const a of args) {
    if (a.startsWith("--mode=")) opts.mode = a.split("=")[1];
    else if (a.startsWith("--codes=")) {
      opts.codes = a.split("=")[1].split(",").map((c) => c.trim()).filter(Boolean);
    }
  }
  // 全量模式: codes=["*"] 等价于"订阅默认 demo 列表"
  if (opts.mode === "all") opts.codes = CONFIG.defaultCodes;
  if (opts.mode === "none") opts.codes = [];
  return opts;
}

// ────────────────────── 工具 ──────────────────────
function ts() {
  return new Date().toLocaleTimeString("zh-CN", { hour12: false });
}

function log(level, msg) {
  const colors = { INFO: "\x1b[36m", OK: "\x1b[32m", WARN: "\x1b[33m", ERR: "\x1b[31m", DIM: "\x1b[90m" };
  const c = colors[level] || "";
  const r = "\x1b[0m";
  console.log(`${c}[${level.padEnd(4)}]${r} [${ts()}] ${msg}`);
}

async function login(baseUrl, username, password) {
  const url = `${baseUrl}/api/auth/login`;
  log("INFO", `POST ${url} (form-data)`);
  // FastAPI OAuth2PasswordRequestForm 走 application/x-www-form-urlencoded
  const formBody = new URLSearchParams({ username, password });
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: formBody,
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`login failed: HTTP ${res.status} ${text}`);
  }
  const body = await res.json();
  if (!body.access_token) throw new Error("login response missing access_token");
  log("OK", `login OK (role=${body.role || "?"}, sub=${body.user_id || "?"})`);
  return body.access_token;
}

function httpToWs(httpUrl) {
  return httpUrl.replace(/^http/, "ws") + "/ws/quote_update";
}

// ────────────────────── 主流程 ──────────────────────
async function main() {
  const opts = parseArgs();
  CONFIG.wsUrl = httpToWs(CONFIG.baseUrl);

  log("INFO", `=== EvTrade 行情订阅 Demo ===`);
  log("INFO", `base URL:    ${CONFIG.baseUrl}`);
  log("INFO", `ws URL:      ${CONFIG.wsUrl}`);
  log("INFO", `mode:        ${opts.mode}`);
  log("INFO", `codes:       [${opts.codes.join(", ")}] (${opts.codes.length} codes)`);
  log("INFO", `collect:     ${CONFIG.collectSeconds}s`);
  console.log("");

  // 1. 登录 (如果 env 没传 JWT)
  let token = process.env.JWT_TOKEN;
  if (!token) {
    token = await login(CONFIG.baseUrl, CONFIG.username, CONFIG.password);
  } else {
    log("INFO", `use JWT_TOKEN from env (skip login)`);
  }

  // 2. 建 WS 连接
  const wsUrlWithToken = `${CONFIG.wsUrl}?token=${encodeURIComponent(token)}`;
  log("INFO", `connecting to WS ...`);
  const ws = new WebSocket(wsUrlWithToken);
  let connected = false;
  let subscribed = false;

  // 统计
  const stats = {
    connectTs: 0,
    subscribeTs: 0,
    ticksByCode: new Map(), // code → [{ts, last_price, count}]
    totalTicks: 0,
    sampleMsgs: [], // 前 5 条原始消息
    pings: 0,
  };

  return new Promise((resolve, reject) => {
    // 收集超时
    const collectTimer = setTimeout(() => {
      log("WARN", `collect timeout (${CONFIG.collectSeconds}s)`);
      finish("timeout");
    }, CONFIG.collectSeconds * 1000);

    function finish(reason) {
      clearTimeout(collectTimer);
      clearInterval(heartbeatTimer);
      printSummary(reason);
      try { ws.close(); } catch {}
      resolve();
    }

    function printSummary(reason) {
      console.log("");
      log("OK", `=== Summary (reason=${reason}) ===`);
      log("INFO", `connected:           ${connected ? "yes" : "no"}`);
      log("INFO", `subscribed:          ${subscribed ? "yes" : "no"}`);
      log("INFO", `total ticks:         ${stats.totalTicks}`);
      log("INFO", `unique codes:        ${stats.ticksByCode.size}`);
      log("INFO", `server pings recv:   ${stats.pings}`);
      if (stats.connectTs && stats.subscribeTs) {
        log("INFO", `connect→subscribe:   ${stats.subscribeTs - stats.connectTs}ms`);
      }
      if (stats.ticksByCode.size > 0) {
        const sorted = [...stats.ticksByCode.entries()].sort((a, b) => b[1].length - a[1].length);
        console.log("");
        console.log("\x1b[36m┌─────────────┬───────┬───────────┬──────────────┐\x1b[0m");
        console.log("\x1b[36m│ stock_code  │ ticks │ last px   │ first ts     │\x1b[0m");
        console.log("\x1b[36m├─────────────┼───────┼───────────┼──────────────┤\x1b[0m");
        for (const [code, ticks] of sorted) {
          const last = ticks[ticks.length - 1].last_price;
          const firstTs = new Date(ticks[0].ts).toLocaleTimeString("zh-CN", { hour12: false });
          console.log(
            `\x1b[36m│\x1b[0m ${code.padEnd(11)} \x1b[36m│\x1b[0m ${String(ticks.length).padStart(5)} \x1b[36m│\x1b[0m ${(last || 0).toFixed(3).padStart(9)} \x1b[36m│\x1b[0m ${firstTs} \x1b[36m│\x1b[0m`
          );
        }
        console.log("\x1b[36m└─────────────┴───────┴───────────┴──────────────┘\x1b[0m");
      } else {
        log("WARN", `no ticks received`);
        log("INFO", `如果这是 mode=none, 这是预期 (验证零订阅静默丢弃)`);
        log("INFO", `如果这是 mode=specific/all, 检查: 1) hqserver 在跑 2) 代码在 tick 数据源里`);
      }
      if (stats.sampleMsgs.length > 0) {
        console.log("");
        log("INFO", `前 ${stats.sampleMsgs.length} 条原始消息 (msg.type 维度):`);
        for (const m of stats.sampleMsgs) {
          const dataStr = JSON.stringify(m.data || m).slice(0, 120);
          console.log(`  ${m.type.padEnd(20)} ${dataStr}${dataStr.length >= 120 ? "..." : ""}`);
        }
      }
    }

    // 心跳日志
    const heartbeatTimer = setInterval(() => {
      const secs = ((Date.now() - stats.connectTs) / 1000).toFixed(0);
      log("DIM", `[tick=${stats.totalTicks} codes=${stats.ticksByCode.size}] running ${secs}s ...`);
    }, CONFIG.heartbeatSec * 1000);

    ws.on("open", () => {
      connected = true;
      stats.connectTs = Date.now();
      log("OK", `WS connected (pid=${process.pid})`);

      // 3. 订阅
      const msg = { type: "subscribe", stock_codes: opts.codes };
      log("INFO", `send subscribe: ${JSON.stringify(msg)}`);
      ws.send(JSON.stringify(msg));
      subscribed = true;
      stats.subscribeTs = Date.now();
    });

    ws.on("message", (raw) => {
      let msg;
      try { msg = JSON.parse(raw.toString()); } catch { return; }

      // 记录原始样本
      if (stats.sampleMsgs.length < 5) stats.sampleMsgs.push(msg);

      if (msg.type === "ping") {
        stats.pings++;
        // 不打印,太刷屏
        return;
      }

      if (msg.type === "quote" && msg.data) {
        const tick = msg.data;
        const code = tick.stock_code || msg.channel || "?";
        const last = tick.last_price;
        stats.totalTicks++;
        if (!stats.ticksByCode.has(code)) stats.ticksByCode.set(code, []);
        const arr = stats.ticksByCode.get(code);
        if (arr.length === 0 || Date.now() - arr[arr.length - 1].ts > 500) {
          // 第一条 或距上一条 >500ms → 记录(避免刷屏)
          arr.push({ ts: Date.now(), last_price: last });
        } else {
          // 静默更新最后一条的 last_price
          arr[arr.length - 1].last_price = last;
        }
        // 实时打印(简化版,避免刷屏)
        log("INFO", `tick ${code} last=${(last || 0).toFixed(3)} bid1=${tick.bid1_price || 0} ask1=${tick.ask1_price || 0}`);
        return;
      }

      if (msg.type === "subscribe_ack") {
        log("OK", `subscribe_ack: ${JSON.stringify(msg).slice(0, 200)}`);
        return;
      }

      // 其他类型
      log("DIM", `recv msg.type=${msg.type}: ${JSON.stringify(msg).slice(0, 100)}`);
    });

    ws.on("error", (err) => {
      log("ERR", `WS error: ${err.message}`);
      finish("error");
    });

    ws.on("close", (code, reason) => {
      log("WARN", `WS closed: code=${code} reason=${reason.toString().slice(0, 50)}`);
      finish("close");
    });

    // Ctrl+C
    process.on("SIGINT", () => {
      log("WARN", `SIGINT received`);
      finish("sigint");
    });
  });
}

main()
  .then(() => {
    log("OK", `demo done`);
    process.exit(0);
  })
  .catch((err) => {
    log("ERR", `fatal: ${err.message}`);
    console.error(err);
    process.exit(1);
  });

File: examples/external_quote_subscriber.py
#!/usr/bin/env python3
"""
EvTrade 行情订阅 Demo —— 外部系统接入示例

功能：
  1) JWT 登录拿到 token
  2) REST 一次性拉最新快照（可跳过,直接 WS 也可）
  3) WS 长连接到 /ws/quote_update,订阅若干股票
  4) 实时接收 quote_update 推送
  5) 心跳 + 自动重连
  6) 优雅退出（unsubscribe + close）

📌 业务背景：
  EvTrade backend 通过 hqserver 收 XtQuant 行情,落地本地 QuoteCache + 每 60s
  批量 UPSERT 到 MySQL quote_snapshots 表。外部应用只要拿到 JWT,就能通过
  这套 subscribe 协议订阅任意股票（限购 200/stock_code/connection）。

📌 用法：
  pip install websockets requests
  python external_quote_subscriber.py
"""
from __future__ import annotations
import asyncio
import json
import logging
import sys
import time
from typing import List

import requests
import websockets

# ──────────────────── 配置 ────────────────────
BASE_URL = "http://127.0.0.1:8000"        # 直连后端调试
# BASE_URL = "https://evtrade.ngx.evdata.top:50443"   # 经 Nginx 反代的公网地址

USERNAME = "admin"
PASSWORD = "admin123"

WATCH_CODES = ["513800.SH", "510300.SH", "600519.SH"]   # 想要订阅的股票
PING_INTERVAL = 25                                       # 比 server 30s 略短
RECONNECT_BACKOFF = 3                                    # 重连间隔(s)


logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
)
log = logging.getLogger("quote_demo")


# ──────────────────── Step 1: 拿 JWT ────────────────────
def login() -> str:
    """
    POST /api/auth/login 拿 access_token。

    协议:application/x-www-form-urlencoded + OAuth2PasswordRequestForm
    不接受 JSON body！username/password 是 form 字段。
    """
    url = f"{BASE_URL}/api/auth/login"
    resp = requests.post(
        url,
        data={"username": USERNAME, "password": PASSWORD},
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        timeout=5,
    )
    resp.raise_for_status()
    data = resp.json()
    log.info("✅ 登录成功,user=%s role=%s", data["user"]["username"], data["user"]["role"])
    return data["access_token"]


# ──────────────────── Step 2 (可选): REST 一次性拉最新快照 ────────────────────
def fetch_snapshots_once(token: str, codes: List[str]) -> dict:
    """
    POST /api/quote/snapshots 一次返所有请求股票的最新一条快照。

    📌 适用场景：UI 首屏加载, 避免 ws 还没 tick 时空白
    📌 注意 body 是 dict 不是 list —— {stock_codes: [...]}
    """
    url = f"{BASE_URL}/api/quote/snapshots"
    resp = requests.post(
        url,
        json={"stock_codes": codes},
        headers={"Authorization": f"Bearer {token}"},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()["snapshots"]   # { "513800.SH": {22 字段}, ... }


# ──────────────────── Step 3: WS 订阅长连接 ────────────────────
async def run_subscriber(token: str, codes: List[str], max_seconds: int = 15):
    """
    流程:
      connect → receive hello (or first push) → 发送 subscribe
      → 收到 subscribe_ack(包含按订阅股票的最后一帧快照)
      → 进入 recv 循环, 收 quote_update / 服务端 ping
      → 每 PING_INTERVALs 主动 send ping 保活
      → max_seconds 到或 Ctrl+C 时发 unsubscribe + close

    📌 重要协议:
    - 入站 ws message = {type: ..., data...} 全部 JSON
    - 出站:            {type: 'subscribe', stock_codes: [...]} | 'unsubscribe' | 'ping'
    - 入站:            {type: 'subscribe_ack', stock_codes, snapshots}
                       {type: 'unsubscribe_ack', stock_codes}
                       {type: 'quote_update', data: {stock_code, last_price, ...}}
                       {type: 'pong', ts}
    """
    # ws URL：HTTP→ws, HTTPS→wss; /ws/quote_update 是行情频道
    ws_base = BASE_URL.replace("https://", "wss://").replace("http://", "ws://")
    url = f"{ws_base}/ws/quote_update?token={token}"

    log.info("🛰  准备连接 %s", url.split("?")[0] + "?token=***")
    async with websockets.connect(url, ping_interval=None) as ws:    # 自己处理 ping
        log.info("🟢 WS 已建立")

        # 3.1 发送 subscribe（订阅多支股票可以一批发,后端会立即返回最后一帧）
        await ws.send(json.dumps({"type": "subscribe", "stock_codes": codes}))
        log.info("📨 已发送 subscribe: %s", codes)

        # 3.2 进入 recv loop
        t0 = time.time()
        last_ping = time.time()
        quote_count = 0
        try:
            while time.time() - t0 < max_seconds:
                # 计算 timeout：取"距离 max_seconds 还剩"和"距离下次必发 ping 还剩"的较小值
                recv_timeout = min(max_seconds - (time.time() - t0), PING_INTERVAL)
                if recv_timeout <= 0:
                    break

                try:
                    raw = await asyncio.wait_for(ws.recv(), timeout=recv_timeout)
                except asyncio.TimeoutError:
                    # 到了主动 ping 的时间
                    await ws.send(json.dumps({"type": "ping"}))
                    last_ping = time.time()
                    log.debug("💗 sent ping")
                    continue

                msg = json.loads(raw)
                mtype = msg.get("type")

                if mtype == "subscribe_ack":
                    snaps = msg.get("snapshots", {}) or {}
                    log.info("✅ subscribe_ack: %d/%d 个 code 有最新快照", len(snaps), len(codes))
                    for code, snap in snaps.items():
                        log.info("   %s: last_price=%s open=%s high=%s low=%s",
                                 code, snap.get("last_price"), snap.get("open_price"),
                                 snap.get("high_price"), snap.get("low_price"))

                elif mtype == "unsubscribe_ack":
                    log.info("✅ unsubscribe_ack: %s", msg.get("stock_codes"))

                elif mtype == "quote_update":
                    data = msg.get("data", {})
                    code = data.get("stock_code", "?")
                    price = data.get("last_price")
                    quote_count += 1
                    # 高频演示：不要每条都 log,按时间窗口聚合
                    if quote_count % 20 == 1:
                        log.info("📈 quote_update[%d]: %s last=%.3f vol=%s",
                                 quote_count, code, price, data.get("volume"))

                elif mtype == "pong":
                    log.debug("💗 pong")

                elif mtype == "ping":
                    # 服务端主动 ping，回应 pong（这里依赖 http 后端似乎没有显式要求，
                    # 因为服务端 endpoint 注释里说"收 ping 就返 pong"，所以双向对称即可）
                    await ws.send(json.dumps({"type": "pong"}))
                    log.debug("💗 resp pong")

                else:
                    log.warning("⚠️ 未知消息 type=%s", mtype)

                # 周期性主动 ping
                if time.time() - last_ping >= PING_INTERVAL:
                    await ws.send(json.dumps({"type": "ping"}))
                    last_ping = time.time()
                    log.debug("💗 sent ping (loop end)")

        except KeyboardInterrupt:
            log.info("⏹ Ctrl+C,准备退出")
        finally:
            # 取消订阅 + 优雅关闭
            try:
                await ws.send(json.dumps({"type": "unsubscribe", "stock_codes": codes}))
                log.info("📤 unsubscribe 已发送")
                # 给服务端一点时间回 ack
                try:
                    await asyncio.wait_for(ws.recv(), timeout=2)
                except asyncio.TimeoutError:
                    pass
            except Exception as e:
                log.warning("unsubscribe 异常:%s", e)

    log.info("🔚 done. 共收 %d 条 quote_update", quote_count)


# ──────────────────── 入口 ────────────────────
async def main():
    token = login()

    # Step 2 (可选)：REST 一次性拉
    log.info("── Step 2: 一次性 REST 拉快照 ──")
    snaps = fetch_snapshots_once(token, WATCH_CODES)
    if snaps:
        for code, s in list(snaps.items())[:2]:
            log.info("REST snap %s: %s", code, s.get("last_price"))
    else:
        log.info("(cache miss 全部,服务冷启动期正常)")

    # Step 3：WS 订阅推送
    log.info("── Step 3: WS 订阅,持续 60s ──")
    await run_subscriber(token, WATCH_CODES, max_seconds=15)

    log.info("👋 all done,退出")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except Exception as e:
        log.exception("FATAL: %s", e)
        sys.exit(1)
